cpp/src/io/parquet/decode_fixed.cu

```diff
@@ -0,0 +1,492 @@
+      } else if (dtype_len == 8) {
+        if (s->dtype_len_in == 4) {
+          // Reading INT32 TIME_MILLIS into 64-bit DURATION_MILLISECONDS
+          // TIME_MILLIS is the only duration type stored as int32:
+          // https://github.com/apache/parquet-format/blob/master/LogicalTypes.md#deprecated-time-convertedtype
+          gpuOutputFast(s, sb, src_pos, static_cast<uint32_t*>(dst));
+        } else if (s->ts_scale) {
+          gpuOutputInt64Timestamp(s, sb, src_pos, static_cast<int64_t*>(dst));
+        } else {
+          gpuOutputFast(s, sb, src_pos, static_cast<uint2*>(dst));
+        }
+      } else if (dtype_len == 4) {
+        gpuOutputFast(s, sb, src_pos, static_cast<uint32_t*>(dst));
+      } else {
+        gpuOutputGeneric(s, sb, src_pos, static_cast<uint8_t*>(dst), dtype_len);
+      }
+    }
+
+    pos += batch_size;
+  }
+}
+
+/**
+ * @brief Kernel for co the column data stored in the pages
+ *
+ * This function will write the page data and the page data's validity to the
+ * output specified in the page's column chunk. If necessary, additional
+ * conversion will be performed to translate from the Parquet datatype to
+ * desired output datatype (ex. 32-bit to 16-bit, string to hash).
+ *
+ * @param pages List of pages
+ * @param chunks List of column chunks
+ * @param min_row Row index to start reading at
+ * @param num_rows Maximum number of rows to read
+ */
+template <typename level_t>
+__global__ void __launch_bounds__(decode_block_size) gpuDecodePageDataFixed(
+  PageInfo* pages, device_span<ColumnChunkDesc const> chunks, size_t min_row, size_t num_rows)
+{
+  __shared__ __align__(16) page_state_s state_g;
+  __shared__ __align__(16) page_state_buffers_s<rolling_buf_size,  // size of nz_idx buffer
+                                                1,                 // unused in this kernel
+                                                1>                 // unused in this kernel
+    state_buffers;
+
+  page_state_s* const s = &state_g;
+  auto* const sb        = &state_buffers;
+  const int page_idx    = blockIdx.x;
+  const int t           = threadIdx.x;
+  PageInfo* pp          = &pages[page_idx];
+
+  if (!(BitAnd(pages[page_idx].kernel_mask, decode_kernel_mask::FIXED_WIDTH_NO_DICT))) { return; }
+
+  // must come after the kernel mask check
+  [[maybe_unused]] null_count_back_copier _{s, t};
+
+  if (!setupLocalPageInfo(s,
+                          pp,
+                          chunks,
+                          min_row,
+                          num_rows,
+                          mask_filter{decode_kernel_mask::FIXED_WIDTH_NO_DICT},
+                          page_processing_stage::DECODE)) {
+    return;
+  }
+
+  // the level stream decoders
+  __shared__ rle_run<level_t> def_runs[rle_run_buffer_size];
+  rle_stream<level_t, decode_block_size, rolling_buf_size> def_decoder{def_runs};
+
+  bool const nullable = s->col.max_level[level_type::DEFINITION] > 0;
```

Comment:
Do we know null counts by the time we get here? Maybe via metadata (v2 headers, page indexes)? It would be nice to skip a RLE decode of a stream of all 1's (esp. given how spark likes to make everything nullable).

Reply:
Followup: I did a quick check with some data that I know has no nulls, but the schema says `OPTIONAL`, so the max def level is 1 (but always 1). Setting `nullable` to false reduces the dict decode time to 6.7ms from 9.8ms for this data set.